File: Goorm/level2/체크카드.cpp
#include <iostream>
#include <vector>
#include <string>
#include <queue>
#include <utility>

using namespace std;

struct BankAccount {
	int balance;
	queue<int> q_reservation;
	
	BankAccount(int balance_);
	~BankAccount() = default;
	
	void deposit(int amount);
	void pay(int amount);
	void reservation(int amount);
	void process_reservations();
};

BankAccount::BankAccount(int balance_) : balance(balance_)
{}

void BankAccount::deposit(int amount) {
	balance += amount;
}

void BankAccount::pay(int amount) {
	if (balance >= amount)
		balance -= amount;
}

void BankAccount::process_reservations() {
	while (!q_reservation.empty()) {
		if(balance >= q_reservation.front()){
			balance -= q_reservation.front();
			q_reservation.pop();
		} else {
			break;
		}
	}
}

void BankAccount::reservation(int amount) {
	if (balance >= amount && q_reservation.empty())
		balance -= amount;
	else
		q_reservation.push(amount);
}

void parse_transaction(BankAccount& account, vector<pair<string, int>> transactions) {
	for (auto & transaction : transactions) {
		string operation = transaction.first;
		int amount = transaction.second;
		if (operation == "deposit")
			account.deposit(amount);
		else if (operation == "pay")
			account.pay(amount);
		else if (operation == "reservation")
			account.reservation(amount);
		account.process_reservations();
	}
	cout << account.balance << endl;
}

int main() {
	int N, M;
	cin >> N >> M;
	
	BankAccount account(N);
	
	vector<pair<string, int>> transactions;
	transactions.clear();
	
	for (int i = 0; i < M; ++i) {
		string operation;
		int amount;
		cin >> operation >> amount;
		transactions.emplace_back(operation, amount);
	}
	
	parse_transaction(account, transactions);
	
	return 0;
}

File: Goorm/level2/체크카드.py
class BankAccount:
    def __init__(self, balance):
        self.balance = balance
        self.reservation_queue = []
    
    def deposit(self, amount):
        self.balance += amount
    
    def pay(self, amount):
        if self.balance >= amount:
            self.balance -= amount
    
    def reservation(self, amount):
        if self.balance >= amount and len(self.reservation_queue) == 0:
            self.balance -= amount
        else:
            self.reservation_queue.append(amount)
    
    def process_reservations(self):
        while len(self.reservation_queue) > 0:
            if self.balance >= self.reservation_queue[0]:
                self.balance -= self.reservation_queue.pop(0)
            else:
                break

def parse_transactions(account, transactions):
    for transaction in transactions:
        operation, amount = transaction
        if operation == "deposit":
            account.deposit(amount)
        elif operation == "pay":
            account.pay(amount)
        elif operation == "reservation":
            account.reservation(amount)
        account.process_reservations()
    
    print(account.balance)


init_balance, transaction_count = map(int, input().split())
account = BankAccount(init_balance)
transactions = []

for _ in range(transaction_count):
    operation, amount = map(str, input().split())
    transactions.append((operation, int(amount)))

parse_transactions(account, transactions)
